backend/app/services/knowledge/knowledge_service.py

```
@@ -160,6 +161,17 @@ class ExternalKnowledgeBaseListFilters:
     limit: int = 50
 
 
+@dataclass(frozen=True)
+class _DirectAccessPermissionContext:
+    """Permission inputs reused within one knowledge-list request."""
+
+    user: User
+    organization_names: frozenset[str]
+    group_roles: dict[str, GroupRole]
+    direct_members: tuple["ResourceMember", ...]
```

Ruff (0.15.21)

[error] Undefined name `ResourceMember`

(F821)